### PR TITLE
Gh3283 expose chain events in contract calls

### PR DESCRIPTION
**Pull request by uwiger**
See issue #3283 

Still WIP

Still some issues with returning tx events in a dry-run REST call. More debugging needed (and then cleaning out a bunch of debug printouts)

### CODE REVIEW

**Comment by uwiger** [comment]
The branch needs to be rebased against the current master.

**Comment by uwiger** [comment]
A slightly expanded `aehttp_contracts_SUITE` now seems to pass.

Illustration:
```
dry_run_txs(Calls) ->
    Txs = [ #{tx => Tx} || {#{tx_encoded := Tx}, _} <- Calls ],
    {ok, 200, #{ <<"results">> := Results }} = dry_run(Txs),
    {ok, 200, #{ <<"results">> := Results
               , <<"tx_events">> := TxEvents }} = dry_run_w_events(Txs),
    ct:log("TxEvents = ~p", [TxEvents]),
    check_dry_calls(Calls, Results).

dry_run_w_events(Txs) ->
    Host = internal_address(),
    http_request(Host, post, "debug/transactions/dry-run", #{ txs => Txs
                                                            , tx_events => true }).

```

When executed, something like this comes out:
```
TxEvents = [#{<<"info">> =>
                  #{<<"amount">> => 21,<<"fee">> => 0,<<"nonce">> => 0,
                    <<"payload">> => <<"ba_dHJhbnNmZXJfdmFsdWXLvS19">>,
                    <<"recipient_id">> =>
                        <<"ak_2SSLEWqB98goygjkp8FwKTxxE7W6LTLsExED1dAyqdeV4SUD3N">>,
                    <<"sender_id">> =>
                        <<"ak_5icD6SkiyhNbJNNuw1Lv9sthVJRMrfjFagQ2Tn3MUHFp7SU8S">>,
                    <<"ttl">> => 0},
              <<"key">> => <<"transfer_value">>,
              <<"kind">> => <<"internal_call_tx">>,
              <<"tx_hash">> =>
                  <<239,191,189,239,191,189,60,219,175,63,239,191,189,239,191,
                    189,103,15,239,191,189,239,191,189,34,239,191,189,239,191,
                    189,87,6,239,191,189,30,239,191,189,47,61,51,239,191,189,
                    106,239,191,189,107,239,191,189,239,191,189,204,129>>,
              <<"type">> => <<"contract_call_tx">>}]
```

The `transfer_value` event is basically a `spend_tx`, but labeled differently, corresponding to the `transfer_value/4` op in [aefa_chain_api](https://github.com/aeternity/aeternity/blob/GH3283-expose_chain_events_in_contract_calls-uw/apps/aefate/src/aefa_chain_api.erl#L281)

File: apps/aefate/src/aefa_chain_api.erl [hunk 1]
```diff
@@ -608,34 +724,50 @@ aens_resolve_from_pstate(NameString, Key, PState) ->
...
     PreclaimDelta = aec_governance:name_claim_preclaim_delta(),
-    Instructions = [ aeprimop:name_claim_op(Pubkey, NameBin, SaltInt, NameFee, PreclaimDelta) ],
+    Instructions = [ aeprimop:name_claim_op(Pubkey, NameBin, SaltInt, NameFee, PreclaimDelta)
+                   , tx_event_op(aens_claim, {Pubkey, NameBin, SaltInt, NameFee,
+                                              ttl(PreclaimDelta, S)})],
     eval_primops(Instructions, S, size_gas([NameBin])).
 
 aens_transfer(FromPubkey, HashBin, ToPubkey, #state{} = S) when ?IS_ONCHAIN(S) ->
-    Instructions = [aeprimop:name_transfer_op(FromPubkey, account, ToPubkey, HashBin)
+    %% We need to lookup the Name for the tx_event op, unfortunately
```
Comment (hanssv) [inline apps/aefate/src/aefa_chain_api.erl:746]: Would it be possible to keep the Name until this part in the call-chain or would that be too ugly? Looking up the name could be costly, or maybe not 🤔

Comment (uwiger) [reply apps/aefate/src/aefa_chain_api.erl]: It turned out that we didn't need the lookup - nor even the name string - at all.

**Comment by uwiger** [comment]
As suggested by @skkw  in [the following forum post](https://forum.aeternity.com/t/active-aeternity-node-maintenance-iris-hard-fork-release-candidate/7365/64?u=uwigeroferlang.chain), tx events are tagged with more Sophia-like `key` values.

Example:
```
*** User 2020-10-13 16:03:30.317 ***
TxEvents = [#{<<"info">> =>
                  #{<<"amount">> => 6000,<<"fee">> => 0,<<"nonce">> => 0,
                    <<"payload">> => <<"ba_Q2hhaW4uc3BlbmRFa4Tl">>,
                    <<"recipient_id">> =>
                        <<"ak_GPbWuS5ke963swAnNngu548KuWDtjGBokcJD4XKq4zCrFUpen">>,
                    <<"sender_id">> =>
                        <<"ak_2cQHqvhGV3Ejkmm6az1TXN7LJgGARcCoRvtV1qhrvi7rZta3Dk">>,
                    <<"ttl">> => 0},
              <<"key">> => <<"Chain.spend">>,
              <<"kind">> => <<"internal_call_tx">>,
              <<"tx_hash">> =>
                  <<239,191,189,239,191,189,93,112,25,239,191,189,17,195,160,
                    239,191,189,239,191,189,86,7,239,191,189,239,191,189,239,
                    191,189,67,28,68,105,89,239,191,189,45,209,144,239,191,189,
                    239,191,189,34,239,191,189,239,191,189>>,
              <<"type">> => <<"contract_call_tx">>},
            #{<<"info">> =>
                  #{<<"amount">> => 6000,<<"fee">> => 0,<<"nonce">> => 0,
                    <<"payload">> => <<"ba_Q2hhaW4uc3BlbmRFa4Tl">>,
                    <<"recipient_id">> =>
                        <<"ak_2cQHqvhGV3Ejkmm6az1TXN7LJgGARcCoRvtV1qhrvi7rZta3Dk">>,
                    <<"sender_id">> =>
                        <<"ak_2S5pY7dke5s2PEpqGKXMA4annN2js8iCsKERgWKdNFrUhAcCh">>,
                    <<"ttl">> => 0},
              <<"key">> => <<"Chain.spend">>,
              <<"kind">> => <<"internal_call_tx">>,
              <<"tx_hash">> =>
                  <<239,191,189,239,191,189,93,112,25,239,191,189,17,195,160,
                    239,191,189,239,191,189,86,7,239,191,189,239,191,189,239,
                    191,189,67,28,68,105,89,239,191,189,45,209,144,239,191,189,
                    239,191,189,34,239,191,189,239,191,189>>,
              <<"type">> => <<"contract_call_tx">>},
            #{<<"info">> =>
                  #{<<"amount">> => 15000,<<"fee">> => 0,<<"nonce">> => 0,
                    <<"payload">> => <<"ba_Q2hhaW4uc3BlbmRFa4Tl">>,
                    <<"recipient_id">> =>
                        <<"ak_GPbWuS5ke963swAnNngu548KuWDtjGBokcJD4XKq4zCrFUpen">>,
                    <<"sender_id">> =>
                        <<"ak_2cQHqvhGV3Ejkmm6az1TXN7LJgGARcCoRvtV1qhrvi7rZta3Dk">>,
                    <<"ttl">> => 0},
              <<"key">> => <<"Chain.spend">>,
              <<"kind">> => <<"internal_call_tx">>,
              <<"tx_hash">> =>
                  <<239,191,189,85,239,191,189,56,28,239,191,189,239,191,189,
                    83,51,239,191,189,60,24,239,191,189,37,127,74,239,191,189,
                    239,191,189,75,6,239,191,189,239,191,189,239,191,189,88,
                    239,191,189,83,57,16,239,191,189,45,239,191,189,239,191,189>>,
              <<"type">> => <<"contract_call_tx">>}]
```

Feedback requested regarding what to call the `transfer_value` event. It is not labeled `Call.amount`

File: apps/aehttp/src/aehttp_helpers.erl [hunk 2]
```diff
@@ -630,6 +650,38 @@ dry_run_result(Type, {ok, CallObj}, Res) when Type =:= contract_call_tx;
...
+	    lager:debug("Couldn't serialize ~p", [E]),
+	    Acc;
+	SerEvent ->
+	    lager:debug("SerEvent = ~p", [SerEvent]),
+	    [SerEvent | Acc]
+    end.
+
+serialize_event({internal_call_tx, Key}, #{ type    := Type
+					  , tx_hash := TxHash
+					  , info    := Tx } = Evt) ->
```
Comment (gorbak25) [inline apps/aehttp/src/aehttp_helpers.erl:666]: formatting

File: apps/aetx/src/aetx_env.erl [hunk 3]
```diff
@@ -56,11 +56,14 @@
         , set_height/2
         , set_payer/2
         , set_signed_tx/2
-        , tx_event/2
+        , tx_event/3
+	, tx_event/4
```
Comment (gorbak25) [inline apps/aetx/src/aetx_env.erl:60]: formatting

**Review by gorbak25 — COMMENTED** [review]
I like this PR :)
After removing some lager:debug dev printouts and discussing what to call the transfer_value event we should merge this.

**Review by gorbak25 — APPROVED** [review]
LGTM

**Review by velzevur — DISMISSED** [review]

**Review by velzevur — APPROVED** [review]

**Review by gorbak25 — APPROVED** [review]

**Comment by nikita-fuchs** [comment]
hey @davidyuk , do you know whether we have the SDK returning/indicating this somewhere ?

**Comment by gorbak25** [comment]
@nikita-fuchs The JS SDK is being maintained on demand - so far I doesn't seem it was exposed in the JS SDK.

**Pull request merged**